indico/modules/events/editing/operations.py

```diff
@@ -0,0 +1,200 @@
+    """
+
+    def __init__(self):
+        super(InvalidEditableState, self).__init__(_('The requested action is not possible on this revision'))
+
+
+def _ensure_latest_revision(revision):
+    if revision != revision.editable.revisions[-1]:
+        raise InvalidEditableState
+
+
+def _ensure_state(revision, initial=None, final=None):
+    if isinstance(initial, InitialRevisionState):
+        initial = {initial}
+    if isinstance(final, FinalRevisionState):
+        final = {final}
+    if initial is not None and revision.initial_state not in initial:
+        raise InvalidEditableState
+    if final is not None and revision.final_state not in final:
+        raise InvalidEditableState
+
+
+def _make_editable_files(editable, files):
+    editable_files = [
+        EditingRevisionFile(file=file, file_type=file_type)
+        for file_type, file_list in files.viewitems()
+        for file in file_list
+    ]
+    for ef in editable_files:
+        ef.file.claim(contrib_id=editable.contribution.id, editable_type=editable.type.name)
+    return editable_files
+
+
+@no_autoflush
+def create_new_editable(contrib, type_, submitter, files):
+    editable = Editable(contribution=contrib, type=type_)
+    revision = EditingRevision(submitter=submitter,
+                               initial_state=InitialRevisionState.ready_for_review,
+                               files=_make_editable_files(editable, files))
+    editable.revisions.append(revision)
+    db.session.flush()
+    logger.info('Editable [%s] created by %s for %s', type_.name, submitter, contrib)
+    return editable
+
+
+def publish_editable_revision(revision):
+    _ensure_latest_revision(revision)
+    revision.editable.published_revision = revision
+    db.session.flush()
+    logger.info('Revision %r marked as published', revision)
+
+
+@no_autoflush
+def review_editable_revision(revision, editor, action, comment, tags, files=None):
+    _ensure_latest_revision(revision)
+    _ensure_state(revision, initial=InitialRevisionState.ready_for_review, final=FinalRevisionState.none)
+    revision.editor = editor
+    revision.comment = comment
+    revision.tags = tags
+    revision.final_state = {
+        EditingReviewAction.accept: FinalRevisionState.accepted,
+        EditingReviewAction.reject: FinalRevisionState.rejected,
+        EditingReviewAction.update: FinalRevisionState.needs_submitter_confirmation,
+        EditingReviewAction.request_update: FinalRevisionState.needs_submitter_changes,
+    }[action]
+
+    db.session.flush()
+    if action == EditingReviewAction.accept:
+        publish_editable_revision(revision)
+    elif action == EditingReviewAction.update:
+        new_revision = EditingRevision(submitter=editor,
+                                       initial_state=InitialRevisionState.needs_submitter_confirmation,
+                                       files=_make_editable_files(revision.editable, files))
```

Comment:
`_make_editable_files` will fail with the default argument for `files`

Reply:
that will never happen because `files` is always specified for the `update` action. anyway, updated the function to handle it

Reply:
Yeah, but then I guess it shouldn't be a keyword argument?